controller/api/destination/server.go

```diff
@@ -338,41 +299,82 @@ func (s *server) GetProfile(dest *pb.GetDestination, stream pb.Destination_GetPr
 	// the context token which sends updates to the secondary listener.  It is
 	// up to the fallbackProfileListener to merge updates from the primary and
 	// secondary listeners and send the appropriate updates to the stream.
-	if dest.GetContextToken() != "" {
-		ctxToken := s.parseContextToken(dest.GetContextToken())
-
+	if token != "" {
+		ctxToken := s.parseContextToken(token)
 		profile, err := profileID(fqn, ctxToken, s.clusterDomain)
 		if err != nil {
-			log.Debugf("Invalid service %s", path)
+			log.Debug("Invalid service")
 			return status.Errorf(codes.InvalidArgument, "invalid profile ID: %s", err)
 		}
-
 		err = s.profiles.Subscribe(profile, primary)
 		if err != nil {
-			log.Warnf("Failed to subscribe to profile %s: %s", path, err)
+			log.Warnf("Failed to subscribe to profile: %s", err)
 			return err
 		}
 		defer s.profiles.Unsubscribe(profile, primary)
 	}
 
 	profile, err := profileID(fqn, contextToken{}, s.clusterDomain)
 	if err != nil {
-		log.Debugf("Invalid service %s", path)
+		log.Debug("Invalid service")
 		return status.Errorf(codes.InvalidArgument, "invalid profile ID: %s", err)
 	}
 	err = s.profiles.Subscribe(profile, secondary)
 	if err != nil {
-		log.Warnf("Failed to subscribe to profile %s: %s", path, err)
+		log.Warnf("Failed to subscribe to profile: %s", err)
 		return err
 	}
 	defer s.profiles.Unsubscribe(profile, secondary)
 
 	select {
 	case <-s.shutdown:
 	case <-stream.Context().Done():
-		log.Debugf("GetProfile(%+v) cancelled", dest)
+		log.Debug("Cancelled")
 	}
+	return nil
+}
 
+func (s *server) translateEndpointProfile(
+	address *watcher.Address,
+	port uint32,
+	stream pb.Destination_GetProfileServer,
+) error {
+	log := s.log
+	if address.Pod != nil {
+		log = log.WithField("ns", address.Pod.Namespace).WithField("pod", address.Pod.Name)
+	} else {
+		log = log.WithField("ip", address.IP)
+	}
+
+	opaquePorts, err := getAnnotatedOpaquePorts(address.Pod, s.defaultOpaquePorts)
+	if err != nil {
+		return fmt.Errorf("failed to get opaque ports for pod: %w", err)
+	}
+	var endpoint *pb.WeightedAddr
+	endpoint, err = s.createEndpoint(*address, opaquePorts)
+	if err != nil {
+		return fmt.Errorf("failed to create endpoint: %w", err)
+	}
+	translator := newEndpointProfileTranslator(address.Pod, port, endpoint, stream, s.log)
```

Comment:
```suggestion
	translator := newEndpointProfileTranslator(address.Pod, port, endpoint, stream, log)
```

Reply:
thanks i'll include that in the follow up